File: src/org/jetbrains/jet/j2k/ast/Function.java
package org.jetbrains.jet.j2k.ast;

import org.jetbrains.annotations.NotNull;
import org.jetbrains.jet.j2k.util.AstUtil;

import java.util.LinkedList;
import java.util.List;
import java.util.Set;

/**
 * @author ignatov
 */
public class Function extends Member {
  private final Identifier myName;
  private final Type myType;
  private final List<Element> myTypeParameters;
  final Element myParams;
  final Block myBlock;

  public Function(Identifier name, Set<String> modifiers, Type type, List<Element> typeParameters, Element params, Block block) {
    myName = name;
    myModifiers = modifiers;
    myType = type;
    myTypeParameters = typeParameters;
    myParams = params;
    myBlock = block;
  }

  private String typeParametersToKotlin() {
    return myTypeParameters.size() > 0 ? "<" + AstUtil.joinNodes(myTypeParameters, COMMA_WITH_SPACE) + ">" : EMPTY;
  }

  private boolean hasWhere() {
    for (Element t : myTypeParameters)
      if (t instanceof TypeParameter && ((TypeParameter) t).hasWhere())
        return true;
    return false;
  }

  private String typeParameterWhereToKotlin() {
    if (hasWhere()) {
      List<String> wheres = new LinkedList<String>();
      for (Element t : myTypeParameters)
        if (t instanceof TypeParameter)
          wheres.add(((TypeParameter) t).getWhereToKotlin());
      return SPACE + "where" + SPACE + AstUtil.join(wheres, COMMA_WITH_SPACE) + SPACE;
    }
    return EMPTY;
  }

  String modifiersToKotlin() {
    List<String> modifierList = new LinkedList<String>();

    if (isAbstract())
      modifierList.add(Modifier.ABSTRACT);

    if (myModifiers.contains(Modifier.OVERRIDE))
      modifierList.add(Modifier.OVERRIDE);

    if (!myModifiers.contains(Modifier.OVERRIDE) && !myModifiers.contains(Modifier.FINAL))
      modifierList.add(Modifier.OPEN);

    modifierList.add(accessModifier());

    if (modifierList.size() > 0)
      return AstUtil.join(modifierList, SPACE) + SPACE;

    return EMPTY;
  }

  @NotNull
  @Override
  public String toKotlin() {
    return modifiersToKotlin() + "fun" + SPACE + myName.toKotlin() + typeParametersToKotlin() + "(" + myParams.toKotlin() + ")" + SPACE + COLON +
      SPACE + myType.toKotlin() + SPACE +
      typeParameterWhereToKotlin() +
      myBlock.toKotlin();
  }
}

File: src/org/jetbrains/jet/j2k/ast/Function.kt
package org.jetbrains.jet.j2k.ast

import java.util.LinkedList
import org.jetbrains.jet.j2k.ast.types.Type

public open class Function(val name : Identifier,
                           val docComments: List<Node>,
                           modifiers : Set<Modifier>,
                           val `type` : Type,
                           val typeParameters : List<Element>,
                           val params : Element,
                           var block : Block?) : Member(modifiers) {
    private fun typeParametersToKotlin() : String {
        return (if (typeParameters.size() > 0)
            "<" + typeParameters.map { it.toKotlin() }.makeString(", ") + ">"
        else
            "")
    }

    private fun hasWhere() : Boolean = typeParameters.any { it is TypeParameter && it.hasWhere() }

    private fun typeParameterWhereToKotlin() : String {
        if (hasWhere())
        {
            val wheres = typeParameters.filter { it is TypeParameter }.map { ((it as TypeParameter).getWhereToKotlin() )}
            return " where " + wheres.makeString(", ") + " "
        }

        return ""
    }

    open fun modifiersToKotlin() : String {
        val modifierList: List<Modifier> = arrayList()
        val accessModifier = accessModifier()
        if (accessModifier != null) {
            modifierList.add(accessModifier)
        }

        if (isAbstract()) {
            modifierList.add(Modifier.ABSTRACT)
        }

        if (modifiers.contains(Modifier.OVERRIDE)) {
            modifierList.add(Modifier.OVERRIDE)
        }

        if (!modifiers.contains(Modifier.ABSTRACT) &&
            !modifiers.contains(Modifier.OVERRIDE) &&
            !modifiers.contains(Modifier.FINAL) &&
            !modifiers.contains(Modifier.PRIVATE)) {
            modifierList.add(Modifier.OPEN)
        }

        if (modifiers.contains(Modifier.NOT_OPEN)) {
            modifierList.remove(Modifier.OPEN)
        }

        return modifierList.toKotlin()
    }

    public override fun toKotlin() : String {
        return docComments.toKotlin("\n", "", "\n") +
               modifiersToKotlin() +
               "fun " + name.toKotlin() +
               typeParametersToKotlin() +
               "(" + params.toKotlin() + ") : " +
               `type`.toKotlin() + " "+ typeParameterWhereToKotlin() +
               block?.toKotlin()
    }
}
